code/datums/logging/_log_entry.dm

```diff
@@ -0,0 +1,177 @@
+	/// The message this log entry contains.
+	var/message
+	/// Whether this log entry is public or not.
+	var/private = FALSE
+	/// The location this log entry was created at.
+	var/list/location
+	/// A lazy list of tags that can be used to filter this log entry.
+	var/list/tags
+
+	// optional settings
+
+	/// The name of the datum that created this log entry.
+	var/source_name
+	/// The ckey, if applicable, of the user that created this log entry.
+	var/source_ckey
+	/// The reference to the source of this log entry; for the default inspection
+	var/datum/weakref/source_weakref
+	/// The name of the target datum, if applicable, of this log entry.
+	var/target_name
+	/// The ckey, if applicable, of the target datum, if applicable, of this log entry.
+	var/target_ckey
+
+	/// extended information about this log entry; associative list
+	var/list/extended_data
+
+/datum/log_entry/New(message)
+	..()
+	SHOULD_CALL_PARENT(TRUE)
+	unix_timestamp = world.realtime
+	world_timestamp = world.time
+	round_id = GLOB.round_id
+	server_name = CONFIG_GET(string/cross_comms_name) || "Unknown"
+	if(!category)
+		CRASH("Log entry created without a category.")
+	if(!message)
+		CRASH("Log entry created without a message.")
+	src.message = message
+	SSlogging.append_entry(src)
+
+/datum/log_entry/proc/as_private()
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	private = TRUE
+	return src
+
+/datum/log_entry/proc/with_location(atom/location)
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if(istype(location, /client)) // heathens
+		var/client/client = location
+		location = client.mob
+	location = list(location.x, location.y, location.z)
+	return src
+
+/datum/log_entry/proc/with_tags(...)
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	LAZYINITLIST(tags)
+	for(var/tag in args)
+		tags += tag
+	return src
+
+/datum/log_entry/proc/with_source(source, append_client_extended_data = TRUE)
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	var/client/client = get_player_client(source)
+	if(client)
+		source_ckey = client.ckey
+		if(append_client_extended_data)
+			with_extended_data("client_mob_name", client.mob.name)
+			with_extended_data("client_mob_type", client.mob.type)
+			with_extended_data("client_is_observer", isobserver(client.mob))
+			with_extended_data("client_is_admin", is_admin(client.mob))
+	source = "[source]"
+	source_weakref = WEAKREF(source)
+	return src
+
+/datum/log_entry/proc/with_target(target)
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	var/client/client = get_player_client(target)
+	if(client)
+		target_ckey = client.ckey
+	target = "[target]"
+	return src
+
+/datum/log_entry/proc/with_extended_data(key, value)
+	RETURN_TYPE(/datum/log_entry)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	LAZYINITLIST(extended_data)
+	extended_data[key] = value
+	return src
+
+/// Construct this log entry into a list, ready for JSON encoding. Add your own data to the list before returning it.
```

Comment:
It seems safer to me to make a separate proc where the user returns a list which is then patched on